src/dx/subscribe/event_engine/types.rs

```diff
@@ -0,0 +1,235 @@
+//! Subscribe event engine module types.
+//!
+//! This module contains the [`SubscribeInput`] type, which represents
+//! user-provided channels and groups for which real-time updates should be
+//! retrieved from the [`PubNub`] network.
+//!
+//! [`PubNub`]:https://www.pubnub.com/
+
+use crate::{
+    core::PubNubError,
+    lib::{
+        alloc::collections::HashSet,
+        core::ops::{Add, AddAssign, Sub, SubAssign},
+    },
+    subscribe::SubscribeCursor,
+};
+
+/// User-provided channels and groups for subscription.
+///
+/// Object contains information about channels and groups for which real-time
+/// updates should be retrieved from the [`PubNub`] network.
+///
+/// [`PubNub`]:https://www.pubnub.com/
+#[derive(Clone, Debug, PartialEq)]
+pub struct SubscribeInput {
+    /// Optional list of channels.
+    ///
+    /// List of channels for which real-time updates should be retrieved
+    /// from the [`PubNub`] network.
+    ///
+    /// List is optional if there is at least one `channel_group` provided.
+    ///
+    /// [`PubNub`]:https://www.pubnub.com/
+    pub channels: Option<HashSet<String>>,
+
+    /// Optional list of channel groups.
+    ///
+    /// List of channel groups for which real-time updates should be retrieved
+    /// from the [`PubNub`] network.
+    ///
+    /// [`PubNub`]:https://www.pubnub.com/
+    pub channel_groups: Option<HashSet<String>>,
+
+    /// Whether user input is empty or not.
+    pub is_empty: bool,
+}
+
+#[allow(dead_code)]
```

Comment:
Why is this allow needed?
Also for merge with presence?

Reply:
It was part of a commit that was not pushed before the code was finished.